### ISSUE TITLE
Feature: Changing Editor to Monaco

### ISSUE THREAD

**Issue by BanceDev** [neocities/neocities#543]
This may have been suggested before and is something that is an intentional choice but in the case that it hasn't been mentioned before I wanted to suggest it. I primarily work on modding/developer tools in my spare time and a big part of the neocities experience is inevitably going to be in its tooling.

I think switching the embedded text editor to monaco would be a great move since we could at minimum have autocomplete out of the box, but we can also bring in language server integrations as well which would be nice for catching some syntax issues in Javascript. It also means that we can offer a much wider selection of themes for the editor since monaco is so widely supported due to its use in vscode.

I'm also entirely willing to implement this feature myself if the interest is there. I've done it before for another website and I think it could be done relatively painlessly for neocities as well. :)

**Comment by BanceDev** [comment]
I made a [test commit](https://github.com/neocities/neocities/compare/master...BanceDev:neocities:monaco-editor) to make a proof of concept for this. I really think that monaco is worth considering. Ace starts lagging at just a couple thousand lines but in my testing I was editing a js file over 20k lines with no lag. It also has autocomplete suggestions which is really nice.

The only thing I think worth noting on the outset is that my test commit here uses the CDN for monaco while it looks like the ace editor code was being served by the neocities host. Obviously I could change it so that we serve monaco instead if pulling code from a CDN is at all problematic. The only other notable loss at the moment is the wider selection of themes. Monaco supports extended themes but I don't have them in this demo yet so it would take some effort to port the old theme options if this editor is considered.

**Comment by kyledrake** [comment]
This is wild news to me.. switching to monaco improves functionality and performance? 

I'm game to try switching and see how many complaints we get if you send a pull request. If Monaco is the VScode default I've been using it for a while anyways.

**Comment by BanceDev** [comment]
Yeah, I saw improved functionality and performance in my limited testing. In particular I saw less latency when dealing with large files. It is the very same Monaco as vscode. I can submit a PR real quick so you can take a look.

**Comment by BanceDev** [comment]
@kyledrake Saw u merged the PR. If you want I can try and tackle the editor themes unless you wanna just do the rest from here.

**Comment by kyledrake** [comment]
I'm almost done with it, wasn't too difficult! I'm hoping to push to production tonight.

**Comment by kyledrake** [comment]
It's been deployed. If there's any issues, please report them here.

**Comment by BanceDev** [comment]
> It's been deployed. If there's any issues, please report them here.

I think it has html error checking on all languages. I'm getting flagging in js for html related issues.

**Comment by kyledrake** [comment]
On it

**Comment by kyledrake** [comment]
@BanceDev I've got a fix in, should be working now. Thanks for catching that.

**Comment by BanceDev** [comment]
Looks good on my end. I'll try and do some more stress testing and see if I can catch anything else. Hopefully we're all good now tho. :)

**Comment by kyledrake** [comment]
I turned off most of the autosuggest stuff because I already got a few complaints about it. I might be able to have some way to turn it on if people want it but this brings it closer to the original formula (tm) so people are more likely to be comfortable with it.

**Comment by BanceDev** [comment]
Yeah. That makes sense. Honestly the biggest reason for the change imo is the performance improvements. Definitely willing to play this one by ear tho and see what people think. Also don't hesitate to ping me if u need help with making changes, I'm happy to help.

**Comment by BanceDev** [comment]
@kyledrake i had someone message me about strangely for html the editor themes we got from that GitHub seem to be the same and the background color just changes. The theming works good for js and css from what I can tell. I also just tested it on a simple one off html page to see if it was the Neocities codebase but it seems that's just how the themes are. We might want to just find a different set to work with.

**Comment by kyledrake** [comment]
The themes came from here:

https://github.com/brijeshb42/monaco-themes

I think they get ported somehow from VSC themes.

If we could figure out how to convert all VS themes, then we could have thousands of themes and people could just choose their own by looking at tons of them. But that seems like a pretty substantial project.

Other things I've noted from complaints but haven't looked at yet:

"new IDE multiselect is alt instead of ctrl"
"middle click to scroll no longer works"

**Comment by 11BelowStudio** [comment]
> The themes came from here:
> 
> https://github.com/brijeshb42/monaco-themes

yeah, about that - https://github.com/brijeshb42/monaco-themes/issues/17 (doesn't work with HTML).

**Comment by kyledrake** [comment]
All of the themes I've tested seem to work fine with HTML. Can you elaborate so I understand?

**Comment by kyledrake** [comment]
This is how it looks on my screen in my browser:

![Screenshot_2024-10-02_13-10-54](https://github.com/user-attachments/assets/5b2857c3-53be-4812-af50-261ebed0d5ab)
![Screenshot_2024-10-02_13-11-32](https://github.com/user-attachments/assets/ad05e83e-704a-4095-8aa4-f4076fc97705)

**Comment by kyledrake** [comment]
Cobalt2 has different font colors and even formatting, so I don't think this is a universal thing.

Anyways, I can revert to the old ace code, I don't really know what else to do at this point. It seems that at least a small number of people are not interested in a more performant editor if it changes the themes, which it sounds like we can't actually port over easily, have bugs or could take a while to do.

**Comment by BanceDev** [comment]
I'd be happy to work on porting the themes. I really would like to see this editor work out :/

**Comment by kyledrake** [comment]
I think I'm going to branch the editor as-is and then revert to the Ace editor. When/if all of the brought up issues get resolved, we can merge it later. I do want to switch, but I just don't have time for the Editor Wars this week, and ultimately there's really not a ton of difference between the two (though the better performance, autocomplete, embedded MDN documentation, and hinting are wonderful features I'd love to have available for people as optional).

Porting from the Ace editor themes to Monaco sounds like an absolutely monumental project to me personally, but if someone wants to go for it, go for it.

**Comment by 11BelowStudio** [comment]
> All of the themes I've tested seem to work fine with HTML. Can you elaborate so I understand?

> Cobalt2 has different font colors and even formatting, so I don't think this is a universal thing.

Basically, just look at the colours of the html tags and such. **All** of the dark themes all have the same dark grey <>=s, dark blue tag names, light blue tag attribute names, and orange tag attribute values.

Yes, the themes do still have some of their distinct elements which they're supposed to have (such as the backgrounds, Cobalt's formatting, some text being italicized, and the comments/non-syntax text are all formatted as intended). However, the colours for the HTML syntax elements in each of these themes have all been overridden by the colours of the vs-dark and vs-light themes. The only two themes unaffected by this are the light/dark high-contrast themes.

This video I recorded cycling through the colour schemes might make it a bit more obvious (and, again, pay attention to how the HTML tags all have the same colour scheme in each theme)

https://github.com/user-attachments/assets/b89ef606-0c9f-4195-a8dd-f6988c552b68

**Comment by Trashpanda115** [comment]
Hey @kyledrake! Thank you for such a timely response on Issue #546
Thank you for explaining why there are now many older themes that are no longer present, Since I don't particularly work on text editors or their themes, I wasn't quite aware of any of this. As for your offer to add in a theme I see as a sufficient replacement, I have found [This Monokai Charcoal High Contrast](https://vscodethemes.com/e/74th.monokai-charcoal-high-contrast/monokai-charcoal?language=javascript) theme that has a similar high contrast visual style to the theme I used before.
I do see your most recent comment, here, mentioning a potential switch back to the recent editor until all mentioned issues are fixed, if ultimately you decide not to switch to Monaco, I do still admire the timely and informative response you have given me and I hope that whatever you decide to do with the editor is something you want to do, and I hope all goes smoothly with either decision.
-Roxanne S.

**Comment by ammon-m** [comment]
It would be nice to be able to configure some of the Monaco editor settings, as is being worked on in #547.
Since some people prefer the extra features and some don't, I think that change would be nice 👍

**Comment by BanceDev** [comment]
[this repo](https://github.com/ChristopherHButler/vscode-themes-in-monaco?tab=readme-ov-file) is the closest thing to being useful that I've found. Long story short we would have to bring in more dependencies since out of the box monaco for some reason has limited support for changing the highlighting for html which is why ppl are having the issues currently being expressed. might end up being the play to roll things back for the time being if ppl don't wanna have to have limited themes while this is worked out.

**Comment by kyledrake** [comment]
It's been reverted for now in master and on production.

Branch is https://github.com/neocities/neocities/tree/monaco for the monaco version.

I'm happy to reconsider if all of the issues can be addressed that were brought up by everyone.

**Comment by ammon-m** [comment]
Is there a way to opt-in to a branch when using the site?
no worries if that's not possible

**Comment by kyledrake** [comment]
Not in production, no. And I really don't want to go in the direction of having to maintain multiple text editors over minute differences.

**Comment by ammon-m** [comment]
Alright, that makes sense. Thanks for the answer!

**Comment by BanceDev** [comment]
Update: the CSS tokenizer is also good to go so my PR is ready. I think the visuals are now more in line with what people were expecting and imo is pretty close to vscode.

**Comment by kyledrake** [comment]
I wanted to propose a possible third way and see what everyone's thoughts were.

Ace actually does have quite a bit of functionality, which we don't currently enable but with some changes could optionally enable for people with a settings popup: https://ace.c9.io/build/kitchen-sink.html

For example check "autocomplete" and "inline preview for autocomplete" and now you get the autocomplete goodies that Monaco had.

We could get pretty darned close to Monaco's fancy features just by allowing users to enable some of these options.

What does evryone think about this approach?

**Comment by BanceDev** [comment]
I think it could be a serviceable solution especially if most people aren't running into performance issues a lot. My biggest issue with Ace is that I have a couple scripts that are just over 1k lines and I get input latency in Ace but not Monaco. 

It's not the end of the world for me so if most ppl just like ace I would like to see a way to engage more features.

**Comment by kyledrake** [comment]
@BanceDev Could you send me the file that's causing your slowdown? Might be helpful to have for my own testing. If you're not seeing it on the demo it's possible this is a bug somewhere too.

**Comment by BanceDev** [comment]
@kyledrake check out PR #551 for some Ace improvements. I also updated the editor to the newest version which is used in the demo and im not rlly having the lag like I used to.

**Comment by kyledrake** [comment]
Going to call this one done, thanks everyone for the work on this!
